### PR TITLE
Search local events

### PR DESCRIPTION
**Pull request by tansaku**
we ended up with lots of cache files in the previous PR #490 so here's a clean start ...
addressing https://www.pivotaltracker.com/story/show/45392619

### CODE REVIEW

**Comment by lizkca** [comment]
hi Sam , i have done some code in this branch, (have make the cucumber pass).
now how should i push these code?

**Comment by tansaku** [comment]
@lizkca you can't push to this branch unless I give you explicit permissions - I can do that, but it would be great to get your git confidence up first.

in the meantime you can push this branch to your local fork and then open a new pull request.  The command for that for you would be 

```
git push origin  45392619_fresh_search_local_events
```
and then you would open a new pull request - want to give it a try?

**Comment by lizkca** [comment]
ok, i will open a new pull request.

**Comment by sigu** [comment]
Were there changes on the closed PR #490  that are not on this PR that need to be considered?

**Comment by lizkca** [comment]
In my opinion, this PR(#558) is some kind of Refactoring.The #490 PR end up with lots of cache files.(Sam ,am i right?) 
and i'd like to do the refactoring after our longest vacation.i think i will open a new PR soon.(because this PR #558 open by Sam,i am not able to push the code).sorry for the delay.

**Comment by tansaku** [comment]
@lizkca no problem - you are correct, this PR was a refactoring of the original PR in order to remove lots of files that should not have been checked in - we also started over to review how to develop this feature ...

**Comment by tansaku** [comment]
@sigu there are more changes needed here, but we couldn't take them from #490 due to the convoluted series of commits involving many un-needed files

**Comment by tansaku** [comment]
thanks for working on this @nisevi - might be a little while before I can look at it - you might also want to try and pair/coordinate with @lizkca on this ...

**Comment by nisevi** [comment]
@tansaku the only thing that is failing now is the search when we tested with cucumber. The functionality is working if you try it manually.

image

**Comment by nisevi** [comment]
@tansaku I fixed the issue I mentioned, the problem was with the `upcomming` scope. Tests are passing now if you wanna take a look :+1:

**Comment by lizkca** [comment]
thanks for working on this @nisevi.

**Comment by tansaku** [comment]
@nisevi when I run this locally and I navigate to the events page I see this warning flash message:

**Comment by nisevi** [comment]
@tansaku hmmmmmm I will take a look :+1: .

**Comment by nisevi** [comment]
@tansaku now the flash message it should display only when a search was performed :+1:  and no results were found.

**Comment by nisevi** [comment]
Hey @sigu what I have done wrong is to copy/paste the code that I found here https://github.com/AgileVentures/LocalSupport/pull/490 without thinking how the code was working, and just thinking oh! this part is what I need, but I didn't realize what you analyzed today :+1:  so thank you for that :+1: !! my fault there copy/pasting things :P

**Comment by sigu** [comment]
@nisevi I have seen that actually it is still searching the whole database and bringing back results with events with either `description` or `title` that matches the query.  There is no limit to the number of results it returns 
```  ruby
where(contains_description?(keyword).or(contains_title?(keyword)))
```

A scope like the one we have for the `upcoming` events

``` ruby 
  scope :upcoming, lambda { |n|
                               where('start_date >= ?', Time.zone.now.to_date)
                              .order('start_date ASC')
                              .limit(n)
                   }
```

So IMO we should define another `search` scope and limit the result set to 10 too

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -4,6 +4,17 @@ class EventsController < ApplicationController
   before_action :logged_in_user, only: [:new, :create]
   before_action :superadmin?, except:[:show, :index]
 
+  def index
+    upcoming = Event.upcoming(10)
+    query = params['q']
+    @events = query.blank? ? upcoming : upcoming.search(query)
```
Comment (sigu) [inline app/controllers/events_controller.rb]: having `upcoming` as a receiver seems not right as the `search` method still searches the **whole database** and gives back **all records** matching the query

File: app/models/event.rb [hunk 2]
```diff
@@ -7,8 +7,8 @@ class Event < ApplicationRecord
   belongs_to :organisation
 
   scope :upcoming, lambda { |n|
-                               where('start_date > ?', DateTime.current.midnight)
-                              .order('created_at DESC')
+                               where('start_date >= ?', Time.zone.now.to_date)
```
Comment (sigu) [inline app/models/event.rb]: IMO this seems to be out of scope for this PR as it does not involve searching for the events, we can merge it in though :smiley:

File: app/models/event.rb [hunk 3]
```diff
@@ -21,6 +21,11 @@ def self.build_by_coordinates(events = nil)
     Location.build_hash(group_by_coordinates(events))
   end
 
+  def self.search(keyword)
```
Comment (sigu) [inline app/models/event.rb]: Refactoring this into a `scope`

File: app/models/event.rb (hunk 2 above)
Comment (nisevi) [reply app/models/event.rb]: @sigu if you try removing this functionality you will realize why I have change it :+1:

**Comment by nisevi** [comment]
@sigu nice work :+1:
